tests/unit/autocomplete/test_indexer.py

```diff
@@ -0,0 +1,180 @@
+# the ``ModelIndexer`` to ensure we're getting the right data.  This has the
+# downsides that we don't verify anything that we're writing to our sqlite3
+# database.  We could be writing inefficient or completely wrong data to our
+# DB, but as last the ``ModelIndexer`` understands how to query the data, we'll
+# never know.  The reason I think this approach makes sense is that I suspect
+# the schema will change as this feature is developed, and it's more important
+# to ensure that we can read the index we generate compared to verifying the
+# contents of the index.  Once this schema has stabilized we could go in and
+# write tests that work at the sqlite3 layer.
+
+class DummyCommand(object):
+    def __init__(self, command_name, subcommand_table=None, arg_table=None):
+        self.name = command_name
+        if subcommand_table is None:
+            subcommand_table = {}
+        self.subcommand_table = subcommand_table
+        if arg_table is None:
+            arg_table = {}
+        self.arg_table = arg_table
+
+
+class DummyArg(object):
+    def __init__(self, name, cli_type_name='string', nargs=None):
+        self.name = name
+        self.cli_type_name = cli_type_name
+        self.nargs = nargs
+
+
+class TestCanRetrieveCommands(unittest.TestCase):
+    def setUp(self):
+        self.session = mock.Mock(spec=Session)
+        self.aws_command = DummyCommand(
+            # The CLIDriver doesn't actually have a ``name`` property,
+            # but it should probably be 'aws'.
+            command_name=None,
+            arg_table={
+                'region': DummyArg('region'),
+                'endpoint-url': DummyArg('endpoint-url'),
+            },
+            subcommand_table={
+                'ec2': DummyCommand(
+                    command_name='ec2',
+                    subcommand_table={
+                        'describe-instances': DummyCommand(
+                            'describe-instances',
+                            arg_table={
+                                'instance-ids': DummyArg('instance-ids'),
+                                'filters': DummyArg(
+                                    'filters', 'list', nargs='+'),
+                            }
+                        ),
+                        'run-instances': DummyCommand('run-instances'),
+                    }
+                ),
+                's3': DummyCommand(
+                    command_name='s3',
+                    subcommand_table={
+                        'list-objects': DummyCommand('list-objects'),
+                        'put-object': DummyCommand('put-object'),
+                    }
+                )
+            }
+        )
+        # Ideally we just use ':memory:', but that's for
+        # a specific sqlite3 connection so we'd have to
+        # update our interfaces to return/accept a db connection.
+        # I'd prefer not to have db connections as part of the
```

Comment:
How about defaulting to an in memory connection if the filename passed is `None`?

Reply:
How does that help though?  That only works if we accept a db connection as an input param which now couples our indexer to a db connection.

Reply:
The `ModelIndexer` is already coupled to `DatabaseConnection` though, so why does that matter? 

Do you mean the `ModelIndex` class taking a filename vs database connection? (Maybe the naming on these classes is a little confusing). Might be worth renaming `ModelIndexer` and `DatabaseConnection` to `SQLiteModelIndexer` and `SQLiteDatabaseConnection`. `ModelIndex` should be defined by the abstract interface:
```python
class ModelIndex(object):
    def command_names(self, lineage):
        pass

    def arg_names(self, lineage, command_name):
        pass

    def get_argument_data(self, lineage, command_name, arg_name):
        pass
```

And we have one concrete implementation of this: `SQLiteModelIndex`. Realistically, the current `ModelIndex` is already tied to a database connection implicitly (which is fine, but maybe the name should reflect that). If we're going to decouple `ModelIndex` from a database connection let's decouple it from IO entirely and not expect a filename either by having the constructor omitted from the interface.

Reply:
That doesn't really answer my question though.  It sounds like you're saying "they're already somewhat coupled so why not couple them even more?".

Unless there's a concrete benefit as to why it make sense to do this, then this seems like unnecessary design at this point.

Reply:
There's two benefits:
1) The interface for a `ModelIndex` is properly defined and isn't mixed with an implementation that's based on a database connection.
2) Allowing an object that represents the database connection to be passed instead decouples your implementation from any specific IO model (file vs memory), but does couple you more to the fact that a database connection is being used. I think this is fine as literally *every* method in the current implementation of the `ModelIndex` object is assuming you're dealing with a database connection. I don't see why you would want to decouple the `SQLiteModelIndex` object from a database connection. If you want to do something radically different create a different object that meets the already defined `ModelIndex` interface.